src/java/org/apache/cassandra/db/virtual/SnapshotsTable.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.cassandra.db.virtual;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import javax.management.openmbean.TabularData;
+import javax.management.openmbean.TabularDataSupport;
+import com.google.common.collect.ImmutableMap;
+
+import org.apache.cassandra.db.SnapshotDetailsTabularData;
+import org.apache.cassandra.db.marshal.BooleanType;
+import org.apache.cassandra.db.marshal.LongType;
+import org.apache.cassandra.db.marshal.UTF8Type;
+import org.apache.cassandra.dht.LocalPartitioner;
+import org.apache.cassandra.schema.TableMetadata;
+import org.apache.cassandra.service.StorageService;
+
+public class SnapshotsTable extends AbstractVirtualTable
+{
+    private static final String SNAPSHOT_NAME = "snapshot_name";
+    private static final String KEYSPACE_NAME = "keyspace_name";
+    private static final String COLUMNFAMILY_NAME = "columnfamily_name";
+    private static final String TRUE_SIZE = "true_size";
+    private static final String SIZE_ON_DISK = "size_on_disk";
+    private static final String CREATE_TIME = "created_at";
+    private static final String EXPIRATION_TIME = "expires_at";
+    private static final String EPHEMERAL = "ephemeral";
+    
+    SnapshotsTable(String keyspace)
+    {
+        super(TableMetadata.builder(keyspace, "snapshots")
+                           .comment("tables snapshots")
+                           .kind(TableMetadata.Kind.VIRTUAL)
+                           .partitioner(new LocalPartitioner(UTF8Type.instance))
+                           .addPartitionKeyColumn(SNAPSHOT_NAME, UTF8Type.instance)
```

Comment:
I would change the key. Partition key would be keyspace and clustering column would be table and name of the snapshot. That way, you can ask:

````
use system_views;
select * from snapshots;
select * from snapshots where keyspace = 'ks1';
select * from snapshots where keyspace = 'ks1' and table = 'tb1';
select * from snapshots where keyspace = 'ks1' and table = 'tb1' and name = "mysnapshot";
````

In practice, I think it is more common to ask questions like "what snapshots this keyspace / table" has? Instead of querying that table by name of the particular snapshot. What if I do not know the name of the snapshot in advance?

Reply:
+1  on what you said

Reply:
hi @Maxwell-Guo , any chance to rework this as discussed above?

Reply:
hi @smiklosovic ，thanks so much. Busy these days so I feel sorry for the delay in replying to the message.
I have already finished the above work.
Besides I hope you can take a look at the reply of naming below.Thanks

Reply:
@pauloricardomg what do you think about column names? Should we follow the naming in nodetool's output? I think we should not be tied by it.

Reply:
Sorry for the delay - have been busy in the past couple of weeks but will try to be faster on next iterations so we can wrap this up soon.

I agree with with @smiklosovic comment about using more succint column names (ie. `keyspace` vs `keyspace_name` and `table` vs `columnfamily_name`), but I think we should use the column name "tag" and not "name" to refer to a snapshot identifier, and keep this as table partition key. We should enable allow filtering on this virtual table (with [CASSANDRA-18271](https://issues.apache.org/jira/browse/CASSANDRA-18271) to allow querying by non partition keys (ie. `keyspace` or `table`). This would allow the following queries:

```sql
use system_views;
select * from snapshots;
select * from snapshots where tag = "mysnapshot" # query by partition key
select * from snapshots where keyspace = 'ks1'; # query by non-partition keys
select * from snapshots where keyspace = 'ks1' and table = 'tb1';
select * from snapshots where keyspace = 'ks1' and table = 'tb1' and tag = "mysnapshot";
```

What do you think?

Reply:
> I think we should use the column name "tag" and not "name" to refer to a snapshot identifier

Even though we have been using `tag` to refer to snapshot ids, this name is a bit misleading since a resource could in theory have multiple tags (but just one ID). @smiklosovic I wonder if it would make sense to just call it snapshot `id` and update all internal usages to refer to snapshots by `id` and not `name/tag` ?

Updated queries with `id` colum name for snapshot identifier:
```sql
use system_views;
select * from snapshots;
select * from snapshots where id = "mysnapshot" # query by partition key
select * from snapshots where keyspace = 'ks1'; # query by non-partition keys
select * from snapshots where keyspace = 'ks1' and table = 'tb1';
select * from snapshots where keyspace = 'ks1' and table = 'tb1' and id = "mysnapshot";
```

Reply:
Hi @pauloricardomg  and @smiklosovic 
I found that the word keyspace is a reserved word for cql 
https://github.com/apache/cassandra/blob/trunk/src/antlr/Parser.g#L1898
So and error will occurs when doing select when specified keyspace = xxx or select keyspace  where xxx
I think that's why other system table using keyspace_name instand of keyspace
https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/db/SystemKeyspace.java#L238

And I am ok if we change snapshot_name from name to tag, And i have already finish this.

Reply:
yes it is very unfortunate that keyspace is a reserved word. It may be workarounded by 'keyspace' (putting that to quotes) but that is quite offputting and not intuitive for users to deal with this ... So next best thing is "keyspace_name".

Or we fix cql grammar to not fail on that.

I dont object using 'id' but doing that internally might be dedicated to a separate ticket for the sake of simplicity in this ticket.

Filtering will be enabled by default on vtables when not explicitly disallowed so nothing to cover there.

Reply:
Thanks @smiklosovic , So what about using keyspace_name and table_name ? Because I found that some of the system tables just using keyspace_name and table_name , although virtual table is not a real system table, it is similar in a sense. 

I think open a new ticket for change the tag/name to id internally is ok, if we all agree on this I can open a ticket for this and finish the task.
@pauloricardomg  WDYT?

Reply:
I think it is fine to have keyspace_name and table_name.

Reply:
Please change the code to include my suggestion I posted above:

_Partition key would be `keyspace_name` and clustering columns would be `table_name` and `id` of the snapshot._

Reply:
> Partition key would be keyspace_name and clustering columns would be table_name and id of the snapshot.

Why do you think they should be primary keys?  I think primary key of snapshot should be `id`, since this is the primary snapshot identifier (ie. different keyspaces cannot have the same snapshot id). It's still possible to query by `keyspace_name` and `table_name` if needed after CASSANDRA-18238.

Reply:
@pauloricardomg different keyspace and table can have same snapshot id, so first time I use id as partition key and keyspace ,table as clustering key to used to distinguish each other .
In my mind ,I think the scope of snapshot name is bigger than keyspace as different keyspace may have same snapshot name, so first time I treate id as partition key.But latter I am +1 on @smiklosovic , for the first perspective of a user may be the keyspace, and they want to know a certain snapshot under the keyspace when writing a cql.
This is what I am thinking.

Reply:
@pauloricardomg 

Regardless of whether we can query without allow filtering after 18238 I think it is still a good practice to act as we would normally do when modeling the schema.

As explained above and what Maxwell just explained, I think keyspace > table > snapshot id.

We can have same snapshot names after all, no? So if keyspace and table are clustering columns, with two snapshots of the same name, we would have this partition:

    snapshotName1 | keyspace1 | table1
    snapshotName1 | keyspace2 | table2

primary key would be `(snapshotName, (keyspaceName, tableName))`

Is not this counter-intuitive to have a partition which is logically coupling different keyspaces and names under the same snapshot name? That does not make sense to me.

Reply:
If so,It seems that there is no need for me to change the code logic ?

Reply:
@pauloricardomg I think we should bring this to ML as we are changing the behavior suddenly and some snapshots will not be possible to do anymore (snapshots with same name).

Reply:
ok I will send a message shortly to the ML to see what others think. In any case we should make the behavior consistent whatever we agree to (either allow `takeSnapshot` to take snapshot with existing id on different table, or restrict `takeMultipleTableSnapshot` to not allow snapshot with existing id).

Reply:
@pauloricardomg  based on the thread in ML it seems like we are ok with what is here. I put some more comments to PR, @Maxwell-Guo , could you please address them? So we might approach the CI.